File: src/parser.rs
//! CSON text parser — hand-written recursive descent.
//!
//! Parses the Crush Semantic Object Notation text format into the
//! canonical `CsonDocument` type from the parent module.

use super::{CsonAnnotation, CsonDocument, CsonKey, CsonNode, CsonValue};
use std::collections::HashMap;

/// A basic MVP parser for CSON text format.
pub struct CsonParser<'a> {
    pub input: &'a str,
    pub pos: usize,
}

impl<'a> CsonParser<'a> {
    pub fn new(input: &'a str) -> Self {
        Self { input, pos: 0 }
    }

    pub fn error(&self, msg: &str) -> String {
        let mut line = 1;
        let mut col = 1;
        for c in self.input[..self.pos].chars() {
            if c == '\n' {
                line += 1;
                col = 1;
            } else {
                col += 1;
            }
        }
        format!("{} at line {}, col {}", msg, line, col)
    }

    pub fn skip_whitespace_and_comments(&mut self) {
        while self.pos < self.input.len() {
            let rest = &self.input[self.pos..];
            if rest.starts_with("//") || rest.starts_with("#") {
                let end = rest.find('\n').unwrap_or(rest.len());
                self.pos += end;
            } else if rest.starts_with(char::is_whitespace) {
                self.pos += rest.chars().next().unwrap().len_utf8();
            } else {
                break;
            }
        }
    }

    fn parse_quoted_string(&mut self) -> Result<String, String> {
        self.pos += 1; // skip '"'
        let mut s = String::new();
        let mut escaped = false;
        while self.pos < self.input.len() {
            let c = self.input[self.pos..].chars().next().unwrap();
            self.pos += c.len_utf8();
            if escaped {
                s.push(match c {
                    'n' => '\n',
                    'r' => '\r',
                    't' => '\t',
                    '"' => '"',
                    '\\' => '\\',
                    _ => c,
                });
                escaped = false;
            } else if c == '\\' {
                escaped = true;
            } else if c == '"' {
                return Ok(s);
            } else {
                s.push(c);
            }
        }
        Err(self.error("Unclosed string"))
    }

    pub fn parse(&mut self) -> Result<CsonDocument, String> {
        let mut root_obj = HashMap::new();
        let mut current_section = None;
        let mut pending_annotations = Vec::new();
        let mut document_version = "1.0".to_string();

        while self.pos < self.input.len() {
            self.skip_whitespace_and_comments();
            if self.pos >= self.input.len() {
                break;
            }

            let rest = &self.input[self.pos..];

            // 1. Annotations
            if rest.starts_with('@') {
                let ann = self.parse_annotation()?;
                if ann.name == "cson" {
                    if let Some(v) = ann.properties.get("version") {
                        document_version = v.clone();
                    } else if let Some(ref arg) = ann.args {
                        document_version = arg.trim_matches('"').to_string();
                    }
                    // `@cson` is reserved for document metadata (SPEC §2.5): it
                    // sets the version and does NOT attach to the next node.
                    continue;
                }
                pending_annotations.push(ann);
                continue;
            }

            // 2. Sections
            if rest.starts_with('[') {
                self.pos += 1;
                let end = self.input[self.pos..]
                    .find(']')
                    .ok_or_else(|| self.error("Unclosed section"))?;
                let section_name = self.input[self.pos..self.pos + end].trim().to_string();
                self.pos += end + 1;
                current_section = Some(section_name);
                continue;
            }

            // 3. Key-Value pairs
            if let Some((key, mut node)) = self.parse_kv_pair()? {
                node.annotations.append(&mut pending_annotations);

                if let Some(ref sec) = current_section {
                    let sec_key = sec.clone();
                    let section_node = root_obj.entry(sec_key).or_insert_with(|| CsonNode {
                        value: CsonValue::Object(HashMap::new()),
                        confidence: None,
                        annotations: vec![],
                    });

                    if let CsonValue::Object(ref mut map) = section_node.value {
                        let k_str = key.to_string();
                        if map.contains_key(&k_str) {
                            return Err(self.error(&format!("Duplicate key: {}", k_str)));
                        }
                        map.insert(k_str, node);
                    }
                } else {
                    let k_str = key.to_string();
                    if root_obj.contains_key(&k_str) {
                        return Err(self.error(&format!("Duplicate key: {}", k_str)));
                    }
                    root_obj.insert(k_str, node);
                }
            } else {
                break;
            }
        }

        Ok(CsonDocument {
            version: document_version,
            root: CsonNode {
                value: CsonValue::Object(root_obj),
                confidence: None,
                annotations: pending_annotations,
            },
        })
    }

    fn parse_annotation(&mut self) -> Result<CsonAnnotation, String> {
        self.pos += 1; // skip '@'
        let name_end = self.input[self.pos..]
            .find(|c: char| !c.is_alphanumeric() && c != '_')
            .unwrap_or(self.input.len() - self.pos);
        let name = self.input[self.pos..self.pos + name_end].to_string();
        self.pos += name_end;

        self.skip_whitespace_and_comments();

        let mut args = None;
        if self.pos < self.input.len() && self.input[self.pos..].starts_with('(') {
            self.pos += 1;
            let end = self.input[self.pos..]
                .find(')')
                .ok_or_else(|| self.error("Unclosed annotation args"))?;
            args = Some(self.input[self.pos..self.pos + end].to_string());
            self.pos += end + 1;
        }

        self.skip_whitespace_and_comments();

        let mut properties = HashMap::new();
        if self.pos < self.input.len() && self.input[self.pos..].starts_with('{') {
            self.pos += 1;
            loop {
                self.skip_whitespace_and_comments();
                if self.pos >= self.input.len() {
                    return Err(self.error("Unclosed annotation properties"));
                }
                if self.input[self.pos..].starts_with('}') {
                    self.pos += 1;
                    break;
                }

                let key_end = self.input[self.pos..]
                    .find(':')
                    .ok_or_else(|| self.error("Missing colon in annotation property"))?;
                let k = self.input[self.pos..self.pos + key_end].trim().to_string();
                self.pos += key_end + 1;

                self.skip_whitespace_and_comments();
                let v = if self.input[self.pos..].starts_with('"') {
                    self.parse_quoted_string()?
                } else {
                    let val_end = self.input[self.pos..]
                        .find([',', '}'])
                        .unwrap_or(self.input.len() - self.pos);
                    let v_str = self.input[self.pos..self.pos + val_end].trim().to_string();
                    self.pos += val_end;
                    v_str
                };
                properties.insert(k, v);

                self.skip_whitespace_and_comments();
                if self.pos < self.input.len() && self.input[self.pos..].starts_with(',') {
                    self.pos += 1;
                }
            }
        }

        Ok(CsonAnnotation {
            name,
            args,
            properties,
        })
    }

    fn parse_kv_pair(&mut self) -> Result<Option<(CsonKey, CsonNode)>, String> {
        self.skip_whitespace_and_comments();
        if self.pos >= self.input.len() {
            return Ok(None);
        }
        if self.input[self.pos..].starts_with('[')
            || self.input[self.pos..].starts_with('@')
            || self.input[self.pos..].starts_with('}')
            || self.input[self.pos..].starts_with(']')
        {
            return Ok(None);
        }

        let mut is_semantic = false;
        if self.input[self.pos..].starts_with('~') {
            is_semantic = true;
            self.pos += 1;
        }

        let key_str;
        if self.input[self.pos..].starts_with('"') {
            key_str = self.parse_quoted_string()?;
        } else {
            let end = self.input[self.pos..]
                .find(':')
                .ok_or_else(|| self.error("Missing colon in kv pair"))?;
            key_str = self.input[self.pos..self.pos + end].trim().to_string();
            self.pos += end;
        }

        let key = if is_semantic {
            CsonKey::Semantic(key_str)
        } else {
            CsonKey::Exact(key_str)
        };

        self.skip_whitespace_and_comments();
        if !self.input[self.pos..].starts_with(':') {
            return Err(self.error("Expected ':' after key"));
        }
        self.pos += 1;

        self.skip_whitespace_and_comments();

        let (value, mut confidence) = self.parse_value()?;

        // SPEC §3: metadata order is FREE — `v ~0.9 @wip` and `v @wip ~0.9` are the
        // same node, so confidence and annotations may interleave. parse_value
        // consumes a confidence directly after the value; one appearing after an
        // annotation is picked up here.
        //
        // Previously this loop took only `@`, so a later `~` was left unconsumed and
        // the next pair's `find(':')` swallowed it — `a: 1 @note ~0.5\nb: 2` parsed
        // to a bogus key "0.5\nb". Silent corruption, not a clean error.
        let mut annotations = Vec::new();
        loop {
            self.skip_whitespace_and_comments();
            if self.pos >= self.input.len() {
                break;
            }
            if self.input[self.pos..].starts_with('@') {
                annotations.push(self.parse_annotation()?);
                continue;
            }
            match self.parse_confidence()? {
                Some(c) => {
                    if confidence.is_some() {
                        return Err(self.error("Duplicate confidence on a node"));
                    }
                    confidence = Some(c);
                }
                None => break,
            }
        }

        Ok(Some((
            key,
            CsonNode {
                value,
                confidence,
                annotations,
            },
        )))
    }

    /// Consume any annotations trailing a value, e.g. `"x" @note @wip { a: "b" }`.
    /// A node is `value confidence? annotation*` (SPEC §3), and that applies to
    /// array elements as much as to object values.
    fn parse_trailing_annotations(&mut self) -> Result<Vec<CsonAnnotation>, String> {
        let mut out = Vec::new();
        loop {
            let save = self.pos;
            self.skip_whitespace_and_comments();
            if self.pos < self.input.len() && self.input[self.pos..].starts_with('@')
                && !self.input[self.pos..].starts_with("@synthesize")
            {
                out.push(self.parse_annotation()?);
            } else {
                self.pos = save;
                return Ok(out);
            }
        }
    }

    fn parse_value(&mut self) -> Result<(CsonValue, Option<f64>), String> {
        self.skip_whitespace_and_comments();
        if self.pos >= self.input.len() {
            return Err(self.error("Unexpected end of input when parsing value"));
        }

        let value;
        if self.input[self.pos..].starts_with('@') {
            let ann = self.parse_annotation()?;
            if ann.name == "synthesize" {
                value = CsonValue::Synthesize(
                    ann.args.unwrap_or_default().trim_matches('"').to_string(),
                );
            } else {
                return Err(self.error("Only @synthesize is supported as a value annotation"));
            }
        } else if self.input[self.pos..].starts_with('"') {
            value = CsonValue::String(self.parse_quoted_string()?);
        } else if self.input[self.pos..].starts_with("true")
            && (self.pos + 4 == self.input.len()
                || !self.input[self.pos + 4..self.pos + 5]
                    .chars()
                    .next()
                    .unwrap()
                    .is_alphanumeric())
        {
            value = CsonValue::Boolean(true);
            self.pos += 4;
        } else if self.input[self.pos..].starts_with("false")
            && (self.pos + 5 == self.input.len()
                || !self.input[self.pos + 5..self.pos + 6]
                    .chars()
                    .next()
                    .unwrap()
                    .is_alphanumeric())
        {
            value = CsonValue::Boolean(false);
            self.pos += 5;
        } else if self.input[self.pos..].starts_with("null")
            && (self.pos + 4 == self.input.len()
                || !self.input[self.pos + 4..self.pos + 5]
                    .chars()
                    .next()
                    .unwrap()
                    .is_alphanumeric())
        {
            value = CsonValue::Null;
            self.pos += 4;
        } else if self.input[self.pos..].starts_with('{') {
            self.pos += 1;
            let mut map = HashMap::new();
            loop {
                self.skip_whitespace_and_comments();
                if self.pos >= self.input.len() {
                    return Err(self.error("Unclosed object"));
                }
                if self.input[self.pos..].starts_with('}') {
                    self.pos += 1;
                    break;
                }
                if let Some((k, v)) = self.parse_kv_pair()? {
                    let k_str = k.to_string();
                    if map.contains_key(&k_str) {
                        return Err(self.error(&format!("Duplicate key: {}", k_str)));
                    }
                    map.insert(k_str, v);
                }
                self.skip_whitespace_and_comments();
                if self.pos < self.input.len() && self.input[self.pos..].starts_with(',') {
                    self.pos += 1;
                }
            }
            value = CsonValue::Object(map);
        } else if self.input[self.pos..].starts_with('[') {
            self.pos += 1;
            let mut arr = Vec::new();
            loop {
                self.skip_whitespace_and_comments();
                if self.pos >= self.input.len() {
                    return Err(self.error("Unclosed array"));
                }
                if self.input[self.pos..].starts_with(']') {
                    self.pos += 1;
                    break;
                }
                let (v, c) = self.parse_value()?;
                let anns = self.parse_trailing_annotations()?;
                arr.push(CsonNode {
                    value: v,
                    confidence: c,
                    annotations: anns,
                });
                self.skip_whitespace_and_comments();
                if self.pos < self.input.len() && self.input[self.pos..].starts_with(',') {
                    self.pos += 1;
                }
            }
            value = CsonValue::Array(arr);
        } else {
            let mut end = 0;
            let rest = &self.input[self.pos..];
            for (i, c) in rest.char_indices() {
                if c == '\n' || c == ',' || c == '}' || c == ']' || c == '~' {
                    break;
                }
                // A bare token ends at whitespace followed by an annotation
                // (`8080 @wip …`) or a comment — not at the whitespace itself
                // (quoted strings aside, bare values never contain spaces).
                if c == ' ' || c == '\t' {
                    let after = rest[i..].trim_start();
                    if after.starts_with('@') || after.starts_with('#') || after.starts_with("//") {
                        break;
                    }
                }
                if rest[i..].starts_with('#') || rest[i..].starts_with("//") {
                    break;
                }
                end += c.len_utf8();
            }
            let raw_str = self.input[self.pos..self.pos + end].trim();
            value = if let Ok(n) = raw_str.parse::<f64>() {
                CsonValue::Number(n)
            } else if raw_str == "true" {
                CsonValue::Boolean(true)
            } else if raw_str == "false" {
                CsonValue::Boolean(false)
            } else if raw_str == "null" {
                CsonValue::Null
            } else {
                CsonValue::String(raw_str.to_string())
            };
            self.pos += end;
        }

        self.skip_whitespace_and_comments();
        let confidence = self.parse_confidence()?;

        Ok((value, confidence))
    }

    /// Consume `~<number>` if present, returning None when the next token is not
    /// a confidence.
    ///
    /// A `~` also begins a semantic KEY (§2.2), so `~"intent"` must be left alone
    /// for the next pair. Only a `~` followed by something other than a quote is
    /// a confidence.
    fn parse_confidence(&mut self) -> Result<Option<f64>, String> {
        if self.pos >= self.input.len() || !self.input[self.pos..].starts_with('~') {
            return Ok(None);
        }
        let after_tilde = &self.input[self.pos + 1..];
        if after_tilde.starts_with('"') || after_tilde.trim_start().starts_with('"') {
            return Ok(None); // a semantic key beginning the next pair
        }
        let at = self.pos;
        self.pos += 1;
        let end = self.input[self.pos..]
            .find(|c: char| c.is_whitespace() || c == ',' || c == '}' || c == ']')
            .unwrap_or(self.input.len() - self.pos);
        let text = &self.input[self.pos..self.pos + end];
        let Ok(c) = text.parse::<f64>() else {
            self.pos = at;
            return Ok(None);
        };
        self.pos += end;
        // SPEC §7: a parser must reject a confidence outside [0.0, 1.0]. This
        // check was missing entirely -- `~1.5` parsed and projected happily.
        if !(0.0..=1.0).contains(&c) {
            return Err(self.error(&format!("Confidence {c} outside [0.0, 1.0]")));
        }
        Ok(Some(c))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_parse() {
        let input = r#"
            name: "test-agent"
            version: 1.0
            active: true
            port: 8080  # the port
            
            ~"routing rules": { 
                priority: 5 ~0.9
                handler: "default"
            }
        "#;
        let mut parser = CsonParser::new(input);
        let doc = parser.parse().unwrap();
        assert_eq!(doc.version, "1.0");

        if let CsonValue::Object(ref map) = doc.root.value {
            assert!(map.contains_key("name"));
            assert!(map.contains_key("version"));
            assert!(map.contains_key("active"));
            assert_eq!(map.get("port").unwrap().value, CsonValue::Number(8080.0));
        } else {
            panic!("Expected Object root");
        }
    }

    #[test]
    fn test_version_parsing() {
        let input = r#"@cson { version: "2.0" }
            key: "value"
        "#;
        let mut parser = CsonParser::new(input);
        let doc = parser.parse().unwrap();
        assert_eq!(doc.version, "2.0");
    }

    #[test]
    fn test_issue_escaped_string() {
        let input = r#"msg: "he said \"hi\"""#;
        let mut parser = CsonParser::new(input);
        let doc = parser.parse().unwrap();
        if let CsonValue::Object(map) = doc.root.value {
            assert_eq!(
                map.get("msg").unwrap().value,
                CsonValue::String("he said \"hi\"".to_string())
            );
        } else {
            panic!();
        }
    }

    #[test]
    fn test_issue_annotation_comma() {
        let input = r#"@module { purpose: "parse a, b, c" }
        key: 1"#;
        let mut parser = CsonParser::new(input);
        let doc = parser.parse().unwrap();
        if let CsonValue::Object(map) = doc.root.value {
            assert_eq!(
                map.get("key").unwrap().annotations[0]
                    .properties
                    .get("purpose")
                    .unwrap(),
                "parse a, b, c"
            );
        } else {
            panic!();
        }
    }
}

File: impl/python/cson/_parser.py
"""CSON parser — a direct implementation of SPEC.md §2-§5.

Hand-written recursive descent over a character cursor. No dependencies.
"""
from __future__ import annotations

from typing import Any, Optional

from ._model import Annotation, CsonError, Document, Node, Synthesize

_BOOL = {"true": True, "false": False}


class _Cursor:
    __slots__ = ("s", "i", "n")

    def __init__(self, s: str):
        self.s, self.i, self.n = s, 0, len(s)

    # position is derived, not tracked, so the cursor stays cheap to rewind
    def pos(self, at: Optional[int] = None) -> tuple:
        at = self.i if at is None else at
        line = self.s.count("\n", 0, at) + 1
        nl = self.s.rfind("\n", 0, at)
        return line, at - nl

    def err(self, msg: str, at: Optional[int] = None) -> CsonError:
        return CsonError(msg, *self.pos(at))

    def eof(self) -> bool:
        return self.i >= self.n

    def peek(self) -> str:
        return self.s[self.i] if self.i < self.n else ""

    def skip_ws(self, newlines: bool = True) -> None:
        """Whitespace is insignificant except in strings; comments run # -> EOL
        and may appear anywhere whitespace may (SPEC §2.1)."""
        while self.i < self.n:
            c = self.s[self.i]
            if c == "#":
                j = self.s.find("\n", self.i)
                self.i = self.n if j < 0 else j
            elif c == "\n":
                if not newlines:
                    return
                self.i += 1
            elif c in " \t\r":
                self.i += 1
            else:
                return


def _parse_string(c: _Cursor) -> str:
    start = c.i
    if c.peek() != '"':
        raise c.err("Expected a string")
    c.i += 1
    out = []
    while True:
        if c.eof():
            raise c.err("Unclosed string", start)
        ch = c.s[c.i]
        if ch == '"':
            c.i += 1
            return "".join(out)
        if ch == "\\":
            c.i += 1
            if c.eof():
                raise c.err("Unclosed string", start)
            e = c.s[c.i]
            out.append({"n": "\n", "t": "\t", "r": "\r", "b": "\b", "f": "\f",
                        '"': '"', "\\": "\\", "/": "/"}.get(e, e))
            if e == "u":
                hexs = c.s[c.i + 1:c.i + 5]
                if len(hexs) < 4:
                    raise c.err("Truncated \\u escape")
                out[-1] = chr(int(hexs, 16))
                c.i += 4
            c.i += 1
        else:
            out.append(ch)
            c.i += 1


def _parse_number(c: _Cursor) -> float:
    start = c.i
    if c.peek() in "+-":
        c.i += 1
    while c.i < c.n and (c.s[c.i].isdigit() or c.s[c.i] in ".eE+-"):
        # stop at an exponent sign that is not part of the number
        if c.s[c.i] in "+-" and c.s[c.i - 1] not in "eE":
            break
        c.i += 1
    text = c.s[start:c.i]
    try:
        # SPEC §2.3: all numbers are IEEE-754 doubles
        return float(text)
    except ValueError:
        raise c.err(f"Invalid number {text!r}", start) from None


def _parse_annotation(c: _Cursor) -> Annotation:
    start = c.i
    c.i += 1                                    # '@'
    j = c.i
    while c.i < c.n and (c.s[c.i].isalnum() or c.s[c.i] in "_-"):
        c.i += 1
    name = c.s[j:c.i]
    if not name:
        raise c.err("Annotation needs a name", start)

    args = None
    save = c.i
    c.skip_ws(newlines=False)
    if c.peek() == "(":
        open_at = c.i
        c.i += 1
        k = c.s.find(")", c.i)
        if k < 0:
            raise c.err("Unclosed annotation arguments", open_at)
        args = c.s[c.i:k].strip().strip('"')
        c.i = k + 1
    else:
        c.i = save

    props: dict = {}
    save = c.i
    c.skip_ws(newlines=False)
    if c.peek() == "{":
        open_at = c.i
        c.i += 1
        while True:
            c.skip_ws()
            if c.eof():
                raise c.err("Unclosed annotation properties", open_at)
            if c.peek() == "}":
                c.i += 1
                break
            k = c.i
            while c.i < c.n and c.s[c.i] not in ":,}":
                c.i += 1
            pname = c.s[k:c.i].strip()
            if c.peek() != ":":
                raise c.err("Malformed annotation property: expected a colon", k)
            c.i += 1
            c.skip_ws()
            props[pname] = _parse_string(c) if c.peek() == '"' else _parse_bare_prop(c)
            c.skip_ws()
            if c.peek() == ",":
                c.i += 1                        # trailing comma permitted (§5)
    else:
        c.i = save
    return Annotation(name, args, props)


def _parse_bare_prop(c: _Cursor) -> str:
    k = c.i
    while c.i < c.n and c.s[c.i] not in ",}\n":
        c.i += 1
    return c.s[k:c.i].strip()


def _parse_value(c: _Cursor) -> Any:
    c.skip_ws()
    ch = c.peek()
    if ch == "":
        raise c.err("Expected a value")
    if ch == '"':
        return _parse_string(c)
    if ch == "{":
        return _parse_object(c)
    if ch == "[":
        return _parse_array(c)
    if ch == "@":
        at = c.i
        if c.s.startswith("@synthesize", c.i):
            c.i += len("@synthesize")
            c.skip_ws()
            if c.peek() != "(":
                raise c.err("@synthesize needs a parenthesised description", at)
            c.i += 1
            c.skip_ws()
            desc = _parse_string(c)
            c.skip_ws()
            if c.peek() != ")":
                raise c.err("Unclosed @synthesize", at)
            c.i += 1
            return Synthesize(desc)
        raise c.err("Unexpected annotation where a value was expected", at)
    for word, val in (("true", True), ("false", False), ("null", None)):
        if c.s.startswith(word, c.i):
            nxt = c.s[c.i + len(word):c.i + len(word) + 1]
            if not (nxt.isalnum() or nxt == "_"):
                c.i += len(word)
                return val
    if ch.isdigit() or ch in "+-.":
        return _parse_number(c)
    raise c.err(f"Unexpected character {ch!r} where a value was expected")


def _parse_node(c: _Cursor, pending: list) -> Node:
    """node := value (ws '~' number)? (ws annotation)*   -- SPEC §3"""
    value = _parse_value(c)
    conf: Optional[float] = None
    anns = list(pending)

    while True:
        save = c.i
        c.skip_ws(newlines=False)
        if c.peek() == "~" and conf is None:
            at = c.i
            c.i += 1
            c.skip_ws(newlines=False)
            conf = _parse_number(c)
            if not (0.0 <= conf <= 1.0):
                raise c.err(f"Confidence {conf} outside [0.0, 1.0]", at)
            continue
        if c.peek() == "@":
            anns.append(_parse_annotation(c))
            continue
        c.i = save
        break
    return Node(value, conf, tuple(anns))


def _parse_object(c: _Cursor) -> dict:
    open_at = c.i
    c.i += 1                                    # '{'
    out: dict = {}
    while True:
        c.skip_ws()
        if c.eof():
            raise c.err("Unclosed object", open_at)
        if c.peek() == "}":
            c.i += 1
            return out
        key, semantic = _parse_key(c)
        if key in out:
            raise c.err(f"Duplicate key {key!r}")
        node = _parse_node(c, [])
        node.semantic = semantic
        out[key] = node
        c.skip_ws()
        if c.peek() == ",":
            c.i += 1                            # trailing comma permitted
        elif c.peek() not in ("}", ""):
            continue


def _parse_array(c: _Cursor) -> list:
    open_at = c.i
    c.i += 1                                    # '['
    out = []
    while True:
        c.skip_ws()
        if c.eof():
            raise c.err("Unclosed array", open_at)
        if c.peek() == "]":
            c.i += 1
            return out
        out.append(_parse_node(c, []))
        c.skip_ws()
        if c.peek() == ",":
            c.i += 1


def _parse_key(c: _Cursor) -> tuple:
    """key := semantic_key | quoted_key | bare_key   -- SPEC §2.2"""
    c.skip_ws()
    semantic = False
    if c.peek() == "~":
        save = c.i
        c.i += 1
        c.skip_ws(newlines=False)
        if c.peek() != '"':
            c.i = save
            raise c.err("A semantic key must be ~ followed by a quoted string")
        key, semantic = _parse_string(c), True
    elif c.peek() == '"':
        key = _parse_string(c)
    else:
        start = c.i
        while c.i < c.n and c.s[c.i] not in ":\n":
            c.i += 1
        key = c.s[start:c.i].strip()
        if not key:
            raise c.err("Expected a key", start)
    c.skip_ws(newlines=False)
    if c.peek() != ":":
        raise c.err(f"Key {key!r} must be followed by a colon")
    c.i += 1
    return key, semantic


def parse(text: str) -> Document:
    """Parse a CSON document. Raises CsonError with position on failure."""
    c = _Cursor(text)
    root: dict = {}
    version = "1.0"
    pending: list = []
    section: Optional[dict] = None

    while True:
        c.skip_ws()
        if c.eof():
            break
        ch = c.peek()

        if ch == "[":                            # section (§2.4)
            open_at = c.i
            c.i += 1
            j = c.s.find("]", c.i)
            nl = c.s.find("\n", c.i)
            if j < 0 or (nl >= 0 and nl < j):
                raise c.err("Unclosed section", open_at)
            name = c.s[c.i:j].strip()
            c.i = j + 1
            section = {}
            if name in root:
                raise c.err(f"Duplicate key {name!r}", open_at)
            root[name] = Node(section)
            continue

        if ch == "@":
            ann = _parse_annotation(c)
            if ann.name == "cson":               # reserved for doc metadata (§2.5)
                version = ann.properties.get("version") or ann.args or version
            else:
                pending.append(ann)              # attaches to the NEXT node (§4.3)
            continue

        key, semantic = _parse_key(c)
        target = section if section is not None else root
        if key in target:
            raise c.err(f"Duplicate key {key!r}")
        node = _parse_node(c, pending)
        node.semantic = semantic
        pending = []
        target[key] = node

    return Document(root, version)
